### PR TITLE
Add Network Detection

### PR DESCRIPTION
**Pull request by sunof-top**

### CODE REVIEW

File: dashboard/pages/networkdetection.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		fclose($file);
+
+		$file = fopen('/var/dashboard/logs/network-detection.log', 'w');
+		fwrite($file, '');
+		fclose($file);
+
+	}
+}
+?>
+<h1>Network Detection</h1>
```
Comment (qinfengling) [inline dashboard/pages/networkdetection.php:20]: @sunof-top `Server Detection` should be more clear than `Network Detection`

File: monitor-scripts/network-detection.sh [hunk 2]
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
+name="network-detection"
+service=$(cat /var/dashboard/services/$name | tr -d '\n')
+
+#检测网络链接畅通
+function network()
```
Comment (qinfengling) [inline monitor-scripts/network-detection.sh:6]: @sunof-top `function check_server_status()` should be better than `function network()` , network() is a noun, we need a verb.

File: monitor-scripts/network-detection.sh [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    #超时时间
+    local timeout=5
+    #目标网站
+    local target=$1
+    #获取响应状态码
+    local ret_code=`curl -I -s --connect-timeout ${timeout} ${target} -w %{http_code} | tail -n1`
+    if [ "x$ret_code" = "x200" ]; then
+	echo "$2 :network connection successful!" >> /var/dashboard/logs/$name.log
+	#网络畅通
+        return 1
```
Comment (qinfengling) [inline monitor-scripts/network-detection.sh:17]: @sunof-top As a Linux script for function, 0 usually stands for success.

File: systemd/network-detection-check.timer [hunk 4]
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=Run the Packet Forwarder status check every 15 seconds.
```
Comment (qinfengling) [inline systemd/network-detection-check.timer:2]: @sunof-top Please fix the typo in `Description`

**Review by qinfengling — COMMENTED** [review]
Thanks for your pull request, please update the pull request with the advice.

File: systemd/network-detection-check.timer [hunk 5]
```diff
@@ -1,5 +1,5 @@
 [Unit]
-Description=Run the Packet Forwarder status check every 15 seconds.
+Description=Run the Server Detection status check every 15 seconds.
```
Comment (qinfengling) [inline systemd/network-detection-check.timer:2]: @sunof-top `Run the Server Detection every 15 seconds` should be good.
and the filename should be rename to 'server-detection.service`

File: monitor-scripts/server-detection.sh [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	    echo "$2 :network connection failed!" >> /var/dashboard/logs/$name.log
+        return 1
+    fi
+    return 0
+}
+
+
+# Fix invalid status when boot finish
+if [[ $service == 'running' ]]; then
+  if [[ ! -f /tmp/dashboard-$name-flag ]]; then
```
Comment (openxzx) [inline monitor-scripts/server-detection.sh:24]: @sunof-top Please keep the code indentation consistent.

File: monitor-scripts/server-detection.sh [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+service=$(cat /var/dashboard/services/$name | tr -d '\n')
+
+
+function check_server_status()
+{
+    local timeout=5
+    local target=$1
+    local ret_code=`curl -I -s --connect-timeout ${timeout} ${target} -w %{http_code} | tail -n1`
+    if [ "x$ret_code" = "x200" ]; then
+	    echo "$2 :network connection successful!" >> /var/dashboard/logs/$name.log
```
Comment (qinfengling) [inline monitor-scripts/server-detection.sh:12]: @sunof-top Please use the same code style

File: monitor-scripts/server-detection.sh [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    echo 'stopped' > /var/dashboard/services/$name
+  fi
+fi
+
+if [[ $service == 'start' ]]; then
+    touch /tmp/dashboard-$name-flag
+    echo 'running' > /var/dashboard/services/$name
+    echo 'Running Network Detection' > /var/dashboard/logs/$name.log
+
+    check_server_status https://raw.githubusercontent.com/Panther-X/PantherDashboard/main/instal Dashboard-Server
```
Comment (qinfengling) [inline monitor-scripts/server-detection.sh:34]: @sunof-top the URL of the server path should be wrong. please double-check it.

File: dashboard/pages/tools.php [hunk 9]
```diff
@@ -46,5 +46,12 @@
 				<span class="button_title">Set Timezone</span>
 			</a>
 		</li>
+
+		<li id="network_detection_button">
+			<a href="/index.php?page=networkdetection" title="Network Detection">
```
Comment (openxzx) [inline dashboard/pages/tools.php]: @sunof-top networkdetection should modify into serverdetection.

File: monitor-scripts/server-detection.sh [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        port=$(echo ${grpc_addr} | sed 's/"//' | sed 's/"//' | sed 's/http:\/\///' | sed 's/:/ /' | awk '{print $2}' );
+
+        nc -z -w 2 $ip $port
+
+        if [[ $? == 0 ]]; then
+            echo "${grpc_addr} is open" >> /var/dashboard/logs/$name.log
+            echo "${ip},${port},open,"
+        else
+            echo "${grpc_addr} timeout" >> /var/dashboard/logs/$name.log
+            echo "${ip},${port},timeout,"
```
Comment (qinfengling) [inline monitor-scripts/server-detection.sh:48]: @sunof-top For the test result, please update with a meaningful sentence. e.g.
```
Dashboard Server can be connected.
Dashboard Server can't be connected, please check your local network setting.
Validator can be connected
Validator can't be connected, please check your local network setting.
```

**Comment by openxzx** [comment]
@sunof-top Please handle your commit according to the function, after we will merge your pull requests，thank you.

**Comment by openxzx** [comment]
@sunof-top We have merged your code, but we make a small change, so you can check the main branch code, thank you.

File: dashboard/pages/networkdetection.php [hunk 11]
```diff
@@ -0,0 +1,29 @@
+<?php
+$info['service'] = file_get_contents(trim('/var/dashboard/services/network-detection'));
+if($_GET['start'])
+{
+	$info['service'] = trim(file_get_contents('/var/dashboard/services/network-detection'));
+	echo $info['service'];
```
[the remark this answers is not in the document]
Comment (nzamhbchi) [reply dashboard/pages/networkdetection.php]: سلام دستگاه من دنی لیست شده لطفا خارج کنید

**Comment by nzamhbchi** [comment]
لطفا دستگاه من ازبن لیست خارج کنید

**Comment by nzamhbchi** [comment]
راهنمایی کنید دستگاه من ازبن لیست خارج بشه

**Comment by nzamhbchi** [comment]
دستگاه من بسته شد چرا؟

**Comment by nzamhbchi** [comment]
My device is locked, help

**Comment by nzamhbchi** [comment]
My device is locked, help

**Comment by nzamhbchi** [comment]
My device is locked, help

**Comment by nzamhbchi** [comment]
دستگاه من بدون دلیل بسته شده بنام sunny char treuse robin

**Pull request closed without merging**